Issue-9402: [functions] Added ImmediateTrigger class and unit tests

david-streamlio opened the pull request:
Fixes #9402

We needed a new type of triggerer for batch sources that is triggered exactly once, e.g. a S3 source that receives CDC notifications via an SQS topic. In this scenario, the discover phase only needs to run once to setup the SQS connection, etc.

Replaces PR: https://github.com/apache/pulsar/pull/9406

jerrypeng commented on pulsar-broker/src/test/java/org/apache/pulsar/io/PulsarFunctionE2ETest.java:
```diff
@@ -1143,6 +1178,145 @@ public void testPulsarSourceStatsWithUrl() throws Exception {
+        assertEquals(m.tags.get("namespace"), String.format("%s/%s", tenant, namespacePortion));
+        assertEquals(m.tags.get("fqfn"), FunctionCommon.getFullyQualifiedName(tenant, namespacePortion, sourceName));
+        assertTrue(m.value > 0.0);
+
```
Please also delete the source at the end of test

jerrypeng commented on pulsar-broker/src/test/java/org/apache/pulsar/io/PulsarFunctionE2ETest.java:
```diff
@@ -1143,6 +1178,145 @@ public void testPulsarSourceStatsWithUrl() throws Exception {
+        assertEquals(m.tags.get("namespace"), String.format("%s/%s", tenant, namespacePortion));
+        assertEquals(m.tags.get("fqfn"), FunctionCommon.getFullyQualifiedName(tenant, namespacePortion, sourceName));
+        assertTrue(m.value > 0.0);
+
```
Since you using a trigger that triggers on once, the number of messages produced by the data generator batch source should be deterministic.  Please create a reader and read from the output topic to make sure you got the correct number of messages.

david-streamlio replied:
The readNext() method of the BatchDataGeneratorSource has a call to Thread.sleep(50; in it, so the number of messages is not determined by how long each individual test runs. Thus it is NOT deterministic.

jerrypeng commented on pulsar-broker/src/test/java/org/apache/pulsar/io/PulsarFunctionE2ETest.java:
```diff
@@ -217,6 +256,9 @@ void setup(Method method) throws Exception {
                 File file = new File(getClass().getClassLoader().getResource("pulsar-io-data-generator.nar").getFile());
                 Files.copy(file.toPath(), new File(connectorsDir.getAbsolutePath() + "/" + file.getName()).toPath());
+                
+                file = new File(getClass().getClassLoader().getResource("pulsar-io-batch-data-generator.nar").getFile());
```
Since you are copying the connector into the "connectors" dir, please also add a test to test submitting the builtin version of the connector.

jerrypeng commented:
@david-streamlio I don't think you have added the ImmediateTrigger class to the PR

david-streamlio commented:
/pulsarbot run-failure-checks

david-streamlio commented:
@sijie @eolivelli   I replaced the previous PR with this one, and wanted to get your approval before proceeding. I have incorporated all of the changes you requested.

eolivelli reviewed (approved):
I am happy with this patch.
 As soon as the other comments have been addressed/answered 

+1
